services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt

```diff
@@ -0,0 +1,128 @@
+import fi.hel.haitaton.hanke.permissions.Kayttooikeustaso
+import io.mockk.checkUnnecessaryStub
+import io.mockk.clearAllMocks
+import org.junit.jupiter.api.AfterEach
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Nested
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+import org.junit.jupiter.api.extension.ExtendWith
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.boot.test.context.SpringBootTest
+import org.springframework.security.test.context.support.WithMockUser
+import org.springframework.test.context.ActiveProfiles
+
+private const val USERNAME = "test7358"
+
+@SpringBootTest
+@ActiveProfiles("test")
+@WithMockUser(USERNAME)
+@ExtendWith(MockFileClientExtension::class)
+class HakemusServiceITest : DatabaseTest() {
+
+    @Autowired private lateinit var hakemusService: HakemusService
+
+    @Autowired private lateinit var applicationRepository: ApplicationRepository
+
+    @Autowired private lateinit var hakemusFactory: HakemusFactory
+    @Autowired private lateinit var hankeFactory: HankeFactory
+
+    @BeforeEach
+    fun clearMocks() {
+        clearAllMocks()
+    }
+
+    @AfterEach
+    fun checkMocks() {
+        checkUnnecessaryStub()
+    }
+
+    @Nested
+    inner class HakemusResponse {
+        @Test
+        fun `when application does not exist should throw`() {
+            assertThat(applicationRepository.findAll()).isEmpty()
+
+            assertThrows<ApplicationNotFoundException> { hakemusService.hakemusResponse(1234) }
+        }
+
+        @Test
+        fun `returns yhteystiedot and yhteyshenkilot if they're present`() {
+            val hanke = hankeFactory.saveMinimal(generated = true)
+            val application =
+                hakemusFactory.builder(USERNAME, hanke).saveWithYhteystiedot {
+                    hakija(kayttooikeustaso = Kayttooikeustaso.KAIKKI_OIKEUDET)
+                    tyonSuorittaja(kayttooikeustaso = Kayttooikeustaso.KAIKKIEN_MUOKKAUS)
+                    rakennuttaja(kayttooikeustaso = Kayttooikeustaso.HAKEMUSASIOINTI)
+                    asianhoitaja()
+                }
+
+            val response = hakemusService.hakemusResponse(application.id!!)
+
+            assertThat(response.applicationData as JohtoselvitysHakemusDataResponse)
+                .hasAllCustomersWithContacts()
+        }
+    }
+
+    private fun Assert<JohtoselvitysHakemusDataResponse>.hasAllCustomersWithContacts() {
+        prop(JohtoselvitysHakemusDataResponse::customerWithContacts).all {
+            prop(CustomerWithContactsResponse::customer).all {
+                prop(CustomerResponse::type).isEqualTo(CustomerType.COMPANY)
+            }
+        }
+        prop(JohtoselvitysHakemusDataResponse::customerWithContacts).all {
+            prop(CustomerWithContactsResponse::contacts).single().all {
+                prop(ContactResponse::orderer).isTrue()
+            }
+        }
```

Comment:
The two props of the `CustomerWithContactsResponse` can be checked in the section. There's no need for `.all {}` if there's just one test.
```suggestion
        prop(JohtoselvitysHakemusDataResponse::customerWithContacts).all {
            prop(CustomerWithContactsResponse::customer)
                .prop(CustomerResponse::type)
                .isEqualTo(CustomerType.COMPANY)
            prop(CustomerWithContactsResponse::contacts)
                .single()
                .prop(ContactResponse::orderer)
                .isTrue()
        }
```

You could also make an assert for checking the two values from a CustomerWithContactsResponse. Something like
```kotlin
    private fun Assert<CustomerWithContactsResponse>.isCompanyCustomerWithOneContact(orderer: Boolean) {
        prop(CustomerWithContactsResponse::customer)
            .prop(CustomerResponse::type)
            .isEqualTo(CustomerType.COMPANY)

        prop(CustomerWithContactsResponse::contacts)
            .single()
            .prop(ContactResponse::orderer)
            .isEqualTo(orderer)
    }

    private fun Assert<JohtoselvitysHakemusDataResponse>.hasAllCustomersWithContacts() {
        prop(JohtoselvitysHakemusDataResponse::customerWithContacts).isCompanyCustomerWithOneContact(true)
        prop(JohtoselvitysHakemusDataResponse::contractorWithContacts).isCompanyCustomerWithOneContact(false)
        prop(JohtoselvitysHakemusDataResponse::propertyDeveloperWithContacts).isNotNull().isCompanyCustomerWithOneContact(false)
        prop(JohtoselvitysHakemusDataResponse::representativeWithContacts).isNotNull().isCompanyCustomerWithOneContact(false)
    }
```

Reply:
So much cleaner, thank you very much!